### PR TITLE
add numpy spacing frontend function

### LINKED ISSUE #16433: spacing

### PR DESCRIPTION
**Pull request by OziomaEunice**
Including comments to provided clarity on the code
- radian
- arctan2
- hypot

closes #16433

### CODE REVIEW

**Review by juliagsy — COMMENTED** [review]
Hello! Thanks for the changes! We actually have a list of open tasks here, where you can then find the todo lists in the repo's GitHub issues. You'll then need to select an available function to work on for the frontends category.

As for the docstrings, they actually don't apply here as this is the frontends, and the functions are identical to their native function, therefore self-explanatory through their native documentation and would be redundant to add docstrings here.

As for the functions you've added/implemented, please first make sure that they are available in the todo list (not marked as completed or ticked in the todo), more information in the open tasks section of the documentation, https://unify.ai/docs/ivy/overview/contributing/open_tasks.html

Thanks!

**Comment by juliagsy** [comment]
@OziomaEunice Thanks for the clarification! Did you have an issue for that? Could you please link it in the PR or comment with the issue? Thanks a lot!

**Comment by OziomaEunice** [comment]
@juliagsy Sure!
The issue was from:
- [ ] Add Mathematical functions to Numpy Frontend #1525

While the ToDo (Sub-task) is:
- [ ] spacing #16433

**Comment by OziomaEunice** [comment]
I made 2 new PR which now contain the 'spacing' function and its test.

**Comment by juliagsy** [comment]
@OziomaEunice Ahh I see, thanks for the clarification! Since both PRs are about the same changes, would you mind to keep only the newer one of them to avoid duplicates? Thanks a lot!

**Comment by OziomaEunice** [comment]
Ok, I've kept this one here, and removed the other duplicate.

File: ivy/functional/frontends/jax/lax/operators.py [hunk 1]
```diff
@@ -626,3 +626,13 @@ def conj(x):
 @to_ivy_arrays_and_back
 def is_finite(x):
     return ivy.isfinite(x)
+
+
+# approx_min_k
```
Comment (juliagsy) [inline ivy/functional/frontends/jax/lax/operators.py:631]: please remove this function as it is not related to the `spacing` function

File: ivy/functional/frontends/numpy/mathematical_functions/floating_point_routines.py [hunk 2]
```diff
@@ -10,7 +10,8 @@
     handle_numpy_casting,
 )
 
-
+# This is the nextafter function.
```
Comment (juliagsy) [inline ivy/functional/frontends/numpy/mathematical_functions/floating_point_routines.py:13]: comment not needed

File: ivy/functional/frontends/numpy/mathematical_functions/floating_point_routines.py [hunk 3]
```diff
@@ -28,4 +29,26 @@ def _nextafter(
     dtype=None,
     subok=True,
 ):
-    return ivy.nextafter(x1, x2, out=out)
+    return ivy.nextafter(x1, x2, out=out) # Return result
+
+
+
+# This is the spacing function.
```
Comment (juliagsy) [inline ivy/functional/frontends/numpy/mathematical_functions/floating_point_routines.py:36]: comment not needed

File: ivy/functional/frontends/numpy/mathematical_functions/floating_point_routines.py [hunk 4]
```diff
@@ -28,4 +29,26 @@ def _nextafter(
...
+    /,
+    out=None,
+    *,
+    where=True,
+    casting="same_kind",
+    order="K",
+    dtype=None,
+    subok=True,
+):
+    return ivy.spacing(x, out=out) # Return result
```
Comment (juliagsy) [inline ivy/functional/frontends/numpy/mathematical_functions/floating_point_routines.py:54]: the spacing function is not yet available in the ivy functional API, please either implement this frontend function through compositions, or in other cases, propose the function to be implemented in the ivy API, more info here: https://unify.ai/docs/ivy/overview/contributing/open_tasks.html#ivy-experimental-api

**Comment by OziomaEunice** [comment]
Do I merge the branch or is it up to you to do it?

Thanks.

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey! Thanks for contributing! I've left some comments below, feel free to ping me if there's any issue, thanks!

File: ivy/functional/frontends/numpy/mathematical_functions/floating_point_routines.py (hunk 4 above)
Comment (OziomaEunice) [reply ivy/functional/frontends/numpy/mathematical_functions/floating_point_routines.py:54]: Thanks @juliagsy !  In place of

`return ivy.spacing(x, out=out)`

I replaced it with    `return ivy.Array(x, out=out)` .

Is that ok?

**Comment by OziomaEunice** [comment]
Updates made:
1. Remove comments from files as you suggested
2. Replaced `return ivy.spacing(x, out=out)`   with `spacing = ivy.subtract(ivy.nextafter(ivy.abs(x), ivy.sign(x)), x)
return spacing`
3. Remove a function in the operations.py file that was not related to the spacing function.

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey! Thanks for the changes so far! If the tests are passing, the approach is fine. Another addition required is in the ivy/frontends/numpy/__init__.py file. Please import the namespaces at the appropriate location.
1. around line 631
2. around line 713

Could you please update your fork/branch so that it is latest as ivy master as well? Thanks a lot!

**Comment by OziomaEunice** [comment]
@juliagsy I've done the updates as requested in the ivy/frontends/numpy/init.py file.

Cheers :)

**Comment by OziomaEunice** [comment]
@juliagsy are you able to solve the conflicting files? The only thing showing that is conflicting is the addition of spacing function in the init.py file.

**Comment by OziomaEunice** [comment]
Hey thanks! Do I commit the branch or will you be able to do it?

**Comment by juliagsy** [comment]
Hey @OziomaEunice ! The imports should be fine now!
I was diving into the implementation as the tests weren't passing, the implementation should be this:
```
if dtype is None:
        dtype = ivy.dtype(x)
    y = ivy.floor(ivy.log2(ivy.abs(x + 1)))
    spacing = ivy.multiply(ivy.finfo(dtype).eps, ivy.pow(2, y))
    if dtype != "float16":
        spacing = ivy.sign(x) * spacing
    return spacing
```

I've tested and they should be passing, but feel free to verify them again, please ping me when the related changes are done! Thanks!

**Comment by juliagsy** [comment]
Oh! And one last thing, you'll need to set these in the `test_frontend_function` arguments:
```
....
        rtol=1e-03,
        atol=1e-03,
....
```

**Comment by OziomaEunice** [comment]
@juliagsy no worries, I'll do that and push the code in few minutes

**Comment by OziomaEunice** [comment]
> Hey @OziomaEunice ! The imports should be fine now! I was diving into the implementation as the tests weren't passing, the implementation should be this:
> 
> ```
> if dtype is None:
>         dtype = ivy.dtype(x)
>     y = ivy.floor(ivy.log2(ivy.abs(x + 1)))
>     spacing = ivy.multiply(ivy.finfo(dtype).eps, ivy.pow(2, y))
>     if dtype != "float16":
>         spacing = ivy.sign(x) * spacing
>     return spacing
> ```
> 
> I've tested and they should be passing, but feel free to verify them again, please ping me when the related changes are done! Thanks!

Also, do I need to put this code in a file? And if so, where?

**Comment by juliagsy** [comment]
@OziomaEunice this should be the complete function body for `spacing` thanks!

**Comment by OziomaEunice** [comment]
> @OziomaEunice this should be the complete function body for `spacing` thanks!

Sorry...again just for clarification, do I add it to the _spacing function?

**Comment by juliagsy** [comment]
@OziomaEunice yep, that's right! to be specific, it should look like this, thanks!:
```
@handle_numpy_out
@handle_numpy_dtype
@to_ivy_arrays_and_back
@handle_numpy_casting
@from_zero_dim_arrays_to_scalar
def _spacing(
    x,
    /,
    out=None,
    *,
    where=True,
    casting="same_kind",
    order="K",
    dtype=None,
    subok=True,
):
    if dtype is None:
        dtype = ivy.dtype(x)
    y = ivy.floor(ivy.log2(ivy.abs(x + 1)))
    spacing = ivy.multiply(ivy.finfo(dtype).eps, ivy.pow(2, y))
    if dtype != "float16":
        spacing = ivy.sign(x) * spacing
    return spacing
```

**Comment by OziomaEunice** [comment]
Thank you! I've update the code now. What remains is this:

'rtol=1e-03,

atol=1e-03,`


But in which lines do I place this in the test_floating_point_routines.py?

**Comment by juliagsy** [comment]
no problem!
the function used in testing, so after the change, the function arguments should look like this:
```
np_frontend_helpers.test_frontend_function(
        input_dtypes=input_dtypes,
        frontend=frontend,
        test_flags=test_flags,
        fn_tree=fn_tree,
        on_device=on_device,
        rtol=1e-03,
        atol=1e-03,
        x=xs[0],
        out=None,
        where=where,
        casting=casting,
        order="K",
        dtype=dtype,
        subok=True,
    )
```

**Comment by OziomaEunice** [comment]
Thank you so much @juliagsy ! I've push the updated code.

**Comment by OziomaEunice** [comment]
@juliagsy Let me know if there's anything else that needs to be done. In any case, the files have been updated.

**Comment by juliagsy** [comment]
@OziomaEunice I'll wait for the final CI test to complete its run, if everything looks good, I'll go ahead and merge it, thanks again!

**Comment by OziomaEunice** [comment]
@juliagsy No problem, thanks again!

**Comment by juliagsy** [comment]
Thanks for contributing!

**Pull request merged**